design/proposal-updateable-specs.md

```diff
@@ -0,0 +1,90 @@
+# Provide a facility to automatically obtain updated specs
+
+## Background
+
+Troubleshoot is limited in that when folks write an update or a new spec for data collection and analysis, that upgrade is not available when Troubleshoot is invoked from within the kots application manager, or from specs stored in a file.  Updating the local file is simple enough, but updating the specs supplied by kots requires an upgrade to kots in order to collect updates to the default spec supplied by kots, and/or an update to the application which is deployed if the spec supplied by the application is changed.  This update requirement delays folks from accessing updated Troubleshoot specs and therefore discourages people from writing new/updated specs.
+
+It is also difficult to justify the tight coupling of Troubleshoot configuration, which is intended to be an entirely client-side concept, and the application manager which is installed to a Kubernetes environment.
+
+## Goals
+
+* Provide the ability to automatically obtain a spec from a source that receives updates, whenever Troubleshoot is run
+* Decouple the provision of default specs from the kots application manager, so that kots upgrades are not required in order to get access to updated specs
+
+## Non Goals
+
+* Maintain the ability for Troubleshoot to run in an airgapped environment
+* maintain compatilbility with existing Troubleshoot specs
+
+## High-Level Design
+
+Add a new field to the Troubleshoot spec definition, which includes a URI used to locate the current spec online.
+
+If the field is populated, Troubleshoot is to attempt to download the additional spec(s) from the source online, and add to the spec already provided.
+
+Add a CLI flag to prevent any attmept to download, for use in airgap environments and/or when we simply do not want to download.
+
+Update the [default spec included with kots](https://github.com/replicatedhq/kots/blob/main/pkg/supportbundle/defaultspec/spec.yaml) to include a URI pointing at [the troubleshoot-specs repo](https://raw.githubusercontent.com/replicatedhq/troubleshoot-specs/main/in-cluster/default.yaml).
+
+URI's provided may be any of the URI types already accepted by Troubleshoot, including web addresses (https://), secrets, or files.
+
+## Detailed Design
+
+Current spec format is like this example:
+
+```yaml
+apiVersion: troubleshoot.sh/v1beta2
+kind: SupportBundle
+metadata:
+  name: default
+spec:
+  collectors:
+    - clusterInfo: {}
+    - clusterResources: {}
+  analyzers:
+    - cephStatus: {}
+    - longhorn: {}
+```
+
+We could add a new type, `specURIs`, which contains a list of URIs from which to obtain specs:
+
+```yaml
+apiVersion: troubleshoot.sh/v1beta2
+kind: SupportBundle
+metadata:
+  name: default
+spec:
+  specURIs:
```

Comment:
specURIs, additionalURIs, all sorts of names could be used.

Reply:
I was thinking we'd have a list, and merge them all together along with the spec provided originally - that way we can have a master spec with a list of remotes, and/or an update URI.  If we agree to replace the spec entirely then we need to determine if the download of the URI is successful, and what to do if not.  If we have multiple that's a problem (what if we get 2/3 of them?), so I think we go to just one URI per spec, and have that entirely replace if downloaded.

Reply:
I think for updating specifically singular has less edge cases and is more predictable what's going to happen. However, once you're dealing with mergable specs I can how it would be fairly clean to have a spec that just references other specs. That sounds like a good feature, although maybe more related to #650 as an additional input field? Something like "additionalSpecs" that takes a list, then people could give that one input to Troubleshoot and it would be the same as giving multiple arguments.

Back to the name, if we're sticking with a single canonical spec for this field. I think the name should drop to just `Uri` or `URI` (I'm unclear on fully capital vs camelcase. The reason I propose that is because the full path to the field as written would be `spec.specURI` which seems like unnecessary repetition.

Reply:
agreed